### PR TITLE
docs: Musicgen melody model card

### PR DESCRIPTION
**Pull request by AshAnand34**
# What does this PR do?

This pull request updates the documentation for the MusicGen Melody model in `docs/source/en/model_doc/musicgen_melody.md`. The changes aim to simplify and enhance the clarity of the documentation by restructuring the content, adding examples, and improving formatting.

#### Overview and Structure
* Reorganized the content to provide a concise overview of the MusicGen Melody model, highlighting its key features and differences from the original MusicGen.
* Simplified the explanation of the model's architecture, breaking it down into three main components: text encoder, MusicGen Melody decoder, and audio decoder.

#### Examples and Usage
* Replaced lengthy code snippets with streamlined examples for generating music using text and audio prompts, including text-only and unconditional generation scenarios.
* Added detailed examples for using tools like Demucs for melody isolation and quantization techniques for memory optimization.

#### Formatting and Accessibility
* Introduced a sidebar navigation and collapsible sections for better readability and user experience.
* Updated links to external resources and added tooltips for key concepts like quantization and guidance scale.

#36979

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@stevhliu

### CODE REVIEW

File: docs/source/en/model_doc/musicgen_melody.md [hunk 1]
```diff
@@ -14,255 +14,170 @@ rendered properly in your Markdown viewer.
...
-
-Transformers supports both mono (1-channel) and stereo (2-channel) variants of MusicGen Melody. The mono channel versions generate a single set of codebooks. The stereo versions generate 2 sets of codebooks, 1 for each channel (left/right), and each set of codebooks is decoded independently through the audio compression model. The audio streams for each channel are combined to give the final stereo output.
+# MusicGen Melody
 
+[MusicGen Melody](https://huggingface.co/papers/2306.05284) is a single-stage, auto-regressive Transformer model designed for high-quality music generation, conditioned on both text and audio prompts. Unlike its predecessor, MusicGen Melody uses the audio prompt as a direct melodic guide, allowing for more precise control over the generated music.
```
Comment (stevhliu) [inline docs/source/en/model_doc/musicgen_melody.md:27]:
```suggestion
[MusicGen Melody](https://huggingface.co/papers/2306.05284) builds on top of the [MusicGen](./musicgen) model by adding a melody-guided generation approach to enable more controllable audio generation. The model is conditioned on both input text and chromagram. A chromagram better captures the harmonic and melodic features of music.

Unlike MusicGen, MusicGen Melody uses the audio prompt as a conditional signal for the generated audio sample and the conditional text and audio signals are concatenated to the decoder's hidden states.
```

File: docs/source/en/model_doc/musicgen_melody.md [hunk 2]
```diff
@@ -14,255 +14,170 @@ rendered properly in your Markdown viewer.
...
 
-#### Audio Conditional Generation
+You can find all the original [MusicGen Melody](https://huggingface.co/models?sort=downloads&search=facebook%2Fmusicgen) checkpoints on the Hugging Face Hub.
```
Comment (stevhliu) [inline docs/source/en/model_doc/musicgen_melody.md:29]:
```suggestion
You can find all the original MusicGen Melody checkpoints under the [AI at Meta](https://huggingface.co/facebook/models?search=musicgen-melody) organization.
```

File: docs/source/en/model_doc/musicgen_melody.md [hunk 3]
```diff
@@ -14,255 +14,170 @@ rendered properly in your Markdown viewer.
...
 
-The model can generate an audio sample conditioned on a text and an audio prompt through use of the [`MusicgenMelodyProcessor`] to pre-process the inputs.
+> [!TIP]
+> Click on the MusicGen Melody models in the right sidebar for more examples of how to apply the model to various music generation tasks.
```
Comment (stevhliu) [inline docs/source/en/model_doc/musicgen_melody.md:32]:
```suggestion
> [!TIP]
> This model was contributed by [ylacombe](https://huggingface.co/ylacombe).
>
> Click on the MusicGen Melody models in the right sidebar for more examples of how to apply the model to various music generation tasks.
```

File: docs/source/en/model_doc/musicgen_melody.md [hunk 4]
```diff
@@ -14,255 +14,170 @@ rendered properly in your Markdown viewer.
...
 
-In the following examples, we load an audio file using the 🤗 Datasets library, which can be pip installed through the command below:
+The example below demonstrates how to generate music conditioned on an audio melody and a text description using the [`AutoModel`] class.
```
Comment (stevhliu) [inline docs/source/en/model_doc/musicgen_melody.md:34]:
```suggestion
The example below demonstrates how to generate music with [`Pipeline`] or the [`AutoModel`] class.
```

File: docs/source/en/model_doc/musicgen_melody.md [hunk 5]
```diff
@@ -14,255 +14,170 @@ rendered properly in your Markdown viewer.
...
-pip install --upgrade pip
-pip install datasets[audio]
-```
+<hfoptions id="usage">
+<hfoption id="Pipeline">
 
-The audio file we are about to use is loaded as follows:
 ```python
->>> from datasets import load_dataset
+from transformers import pipeline
```
Comment (stevhliu) [inline docs/source/en/model_doc/musicgen_melody.md:40]:
```py
import torch
from transformers import pipeline

pipeline = pipeline("text-to-audio", model="facebook/musicgen-melody", device=0, torch_dtype="auto")
pipeline("80s pop track with bassy drums and synth")
```

File: docs/source/en/model_doc/musicgen_melody.md [hunk 6]
```diff
@@ -14,255 +14,170 @@ rendered properly in your Markdown viewer.
...
+# This pipeline is for text-to-audio, melody conditioning is best done with the AutoModel approach
+pipe = pipeline("text-to-audio", model="facebook/musicgen-melody")
+audio = pipe("80s pop track with bassy drums and synth")
 ```
 
-The audio prompt should ideally be free of the low-frequency signals usually produced by instruments such as drums and bass. The [Demucs](https://github.com/adefossez/demucs/tree/main) model can be used to separate vocals and other signals from the drums and bass components.
-
-If you wish to use Demucs, you first need to follow the installation steps [here](https://github.com/adefossez/demucs/tree/main?tab=readme-ov-file#for-musicians) before using the following snippet:
+</hfoption>
+<hfoption id="AutoModel">
```
Comment (stevhliu) [inline docs/source/en/model_doc/musicgen_melody.md:48]:
```py
import torch
from transformers import AutoProcessor, MusicgenMelodyForConditionalGeneration
from datasets import load_dataset
from demucs import pretrained
from demucs.apply import apply_model
from demucs.audio import convert_audio
from IPython.display import Audio

dataset = load_dataset("sanchit-gandhi/gtzan", split="train", streaming=True)
sample = next(iter(dataset))["audio"]

wav = torch.tensor(sample["array"]).to(torch.float32)

demucs = pretrained.get_model('htdemucs')

wav = convert_audio(wav[None], sample["sampling_rate"], demucs.samplerate, demucs.audio_channels)
wav = apply_model(demucs, wav[None])

processor = AutoProcessor.from_pretrained("facebook/musicgen-melody")
model = MusicgenMelodyForConditionalGeneration.from_pretrained("facebook/musicgen-melody", torch_dtype="auto", device_map="auto")

dataset = load_dataset("sanchit-gandhi/gtzan", split="train", streaming=True)
sample = next(iter(dataset))["audio"]

inputs = processor(
    audio=wav,
    sampling_rate=demucs.samplerate,
    text=["80s blues track with groovy saxophone"],
    padding=True,
    return_tensors="pt",
)
audio_values = model.generate(**inputs, do_sample=True, guidance_scale=3, max_new_tokens=256)

sampling_rate = model.config.audio_encoder.sampling_rate
Audio(audio_values[0].numpy(), rate=sampling_rate)
```

Comment (stevhliu) [reply docs/source/en/model_doc/musicgen_melody.md:48]: Add some text here that says:

```md
Use the [Demucs](https://github.com/adefossez/demucs/tree/main) library to separate vocals and other signals from the drums and bass components for better results.
```

File: docs/source/en/model_doc/musicgen_melody.md [hunk 7]
```diff
@@ -14,255 +14,170 @@ rendered properly in your Markdown viewer.
...
-...     text=["80s blues track with groovy saxophone"],
-...     padding=True,
-...     return_tensors="pt",
-... )
->>> audio_values = model.generate(**inputs, do_sample=True, guidance_scale=3, max_new_tokens=256)
-```
+Quantization reduces the memory burden of large models by representing the weights in a lower precision. Refer to the [Quantization](../quantization/overview) overview for more available quantization backends.
 
-You can also pass the audio signal directly without using Demucs, although the quality of the generation will probably be degraded:
+The example below uses [bitsandbytes quantization](https://huggingface.co/docs/transformers/main/en/main_classes/quantization) to quantize the weights to 8-bit:
```
Comment (stevhliu) [inline docs/source/en/model_doc/musicgen_melody.md:78]:
```suggestion
The example below uses [bitsandbytes](../quantization/bitsandbytes) to only quantize the weights to 8-bit precision.
```

File: docs/source/en/model_doc/musicgen_melody.md [hunk 8]
```diff
@@ -14,255 +14,170 @@ rendered properly in your Markdown viewer.
...
-
-sampling_rate = model.config.audio_encoder.sampling_rate
-Audio(audio_values[0].numpy(), rate=sampling_rate)
-```
-
-Or save them as a `.wav` file using a third-party library, e.g. `soundfile`:
-
-```python
->>> import soundfile as sf
+from transformers import MusicgenMelodyForConditionalGeneration
```
Comment (stevhliu) [inline docs/source/en/model_doc/musicgen_melody.md:81]: Complete this as an end-to-end example

```suggestion
from transformers import AutoProcessor, MusicgenMelodyForConditionalGeneration, BitsAndBytesConfig

bnb_config = BitsAndBytesConfig(
    load_in_8bit=True,
)
```

File: docs/source/en/model_doc/musicgen_melody.md [hunk 9]
```diff
@@ -14,255 +14,170 @@ rendered properly in your Markdown viewer.
...
+import torch
 
->>> sampling_rate = model.config.audio_encoder.sampling_rate
->>> sf.write("musicgen_out.wav", audio_values[0].T.numpy(), sampling_rate)
+model = MusicgenMelodyForConditionalGeneration.from_pretrained(
+    "facebook/musicgen-melody",
+    load_in_8bit=True,
```
Comment (stevhliu) [inline docs/source/en/model_doc/musicgen_melody.md:86]:
```suggestion
    quantization_config=bnb_config,
```

File: docs/source/en/model_doc/musicgen_melody.md [hunk 10]
```diff
@@ -14,255 +14,170 @@ rendered properly in your Markdown viewer.
...
+    device_map="auto"
+)
 ```
 
+## Notes
 
-### Text-only Conditional Generation
+### Differences with MusicGen
+- The audio prompt serves as a conditional signal for the melody, whereas in the original MusicGen, it's used for audio continuation.
+- Conditional text and audio signals are concatenated to the decoder's hidden states, not used as a cross-attention signal.
```
Comment (stevhliu) [inline docs/source/en/model_doc/musicgen_melody.md:95]:
```suggestion
```

File: docs/source/en/model_doc/musicgen_melody.md [hunk 11]
```diff
@@ -14,255 +14,170 @@ rendered properly in your Markdown viewer.
...
 
-The same [`MusicgenMelodyProcessor`] can be used to pre-process a text-only prompt. 
+### Audio Prompting with Demucs
```
Comment (stevhliu) [inline docs/source/en/model_doc/musicgen_melody.md:97]: Remove this section as well since we are already showing it in our examples above for best results

File: docs/source/en/model_doc/musicgen_melody.md [hunk 12]
```diff
@@ -14,255 +14,170 @@ rendered properly in your Markdown viewer.
...
+wav = convert_audio(wav[None], sample_rate, demucs_model.samplerate, demucs_model.audio_channels)
+separated_audio = apply_model(demucs_model, wav[None])
+melody_prompt = separated_audio[0, demucs_model.sources.index('vocals')] # or other source
 ```
 
-The `guidance_scale` is used in classifier free guidance (CFG), setting the weighting between the conditional logits (which are predicted from the text prompts) and the unconditional logits (which are predicted from an unconditional or 'null' prompt). Higher guidance scale encourages the model to generate samples that are more closely linked to the input prompt, usually at the expense of poorer audio quality. CFG is enabled by setting `guidance_scale > 1`. For best results, use `guidance_scale=3` (default).
-
-
-You can also generate in batch:
+### Other Generation Examples
```
Comment (stevhliu) [inline docs/source/en/model_doc/musicgen_melody.md:115]:
```suggestion
```

File: docs/source/en/model_doc/musicgen_melody.md [hunk 13]
```diff
@@ -14,255 +14,170 @@ rendered properly in your Markdown viewer.
...
 
+**Text-only Conditional Generation**
```
Comment (stevhliu) [inline docs/source/en/model_doc/musicgen_melody.md:117]: Indent all the code examples under each list item in the `## Notes` section

```suggestion
- The example below demonstrates text-only conditional generation.
```

File: docs/source/en/model_doc/musicgen_melody.md [hunk 14]
```diff
@@ -14,255 +14,170 @@ rendered properly in your Markdown viewer.
...
-The inputs for unconditional (or 'null') generation can be obtained through the method [`MusicgenMelodyProcessor.get_unconditional_inputs`]:
+inputs = processor(
+    text=["80s pop track with bassy drums and synth", "90s rock song with loud guitars and heavy drums"],
+    padding=True,
+    return_tensors="pt",
+)
+audio_values = model.generate(**inputs, do_sample=True, guidance_scale=3, max_new_tokens=256)
+```
 
+**Unconditional Generation**
```
Comment (stevhliu) [inline docs/source/en/model_doc/musicgen_melody.md:132]:
```suggestion
- The example below demonstrates unconditional generation.
```

File: docs/source/en/model_doc/musicgen_melody.md [hunk 15]
```diff
@@ -14,255 +14,170 @@ rendered properly in your Markdown viewer.
...
+unconditional_inputs = processor.get_unconditional_inputs(num_samples=1)
+audio_values = model.generate(**unconditional_inputs, do_sample=True, max_new_tokens=256)
 ```
 
-### Generation Configuration
-
-The default parameters that control the generation process, such as sampling, guidance scale and number of generated tokens, can be found in the model's generation config, and updated as desired:
-
+**Generation Configuration**
+You can inspect and update the model's generation configuration.
```
Comment (stevhliu) [inline docs/source/en/model_doc/musicgen_melody.md:143]:
```suggestion
- The generation config stores the default parameters that control the generation process such as sampling, guidance scale, and number of generated tokens.

   Any arguments passed to the [`~GenerationMixin.generate`] method supersedes the parameters in the generation config.
```

File: docs/source/en/model_doc/musicgen_melody.md [hunk 16]
```diff
@@ -14,255 +14,170 @@ rendered properly in your Markdown viewer.
...
+# Update config
+model.generation_config.guidance_scale = 4.0
+model.generation_config.max_length = 256
 ```
 
-Note that any arguments passed to the generate method will **supersede** those in the generation config, so setting `do_sample=False` in the call to generate will supersede the setting of `model.generation_config.do_sample` in the generation config.
+### Other Information
+- **Checkpoint Conversion**: Convert original checkpoints using the script at `src/transformers/models/musicgen_melody/convert_musicgen_melody_transformers.py`.
+- **`head_mask`**: The `head_mask` argument is only effective with `attn_implementation="eager"`.
+- **Sampling**: For best results, use sampling (`do_sample=True`).
```
Comment (stevhliu) [inline docs/source/en/model_doc/musicgen_melody.md:158]:
```suggestion
- The `head_mask` argument is only effective with `attn_implementation="eager"`.
- For best results,  set `do_sample=True`.
```

File: docs/source/en/model_doc/musicgen_melody.md [hunk 17]
```diff
@@ -14,255 +14,170 @@ rendered properly in your Markdown viewer.
...
 
 ## Model Structure
```
Comment (stevhliu) [inline docs/source/en/model_doc/musicgen_melody.md:160]: Remove this section and replace with the below. Remember each code snippet should be indented under its list item

```suggestion
- [`MusicgenMelodyForCausalLM`] can be used as a standalone decoder model. Load it by specifying the correct config or accessing it through the `.decoder` attribute of [`MusicgenMelodyForConditionalGeneration`].

   [`MusicgenMelodyForConditionalGeneration`] can be used as a composite model that includes the text and audio encoder.
```

File: docs/source/en/model_doc/musicgen_melody.md [hunk 18]
```diff
@@ -14,255 +14,170 @@ rendered properly in your Markdown viewer.
...
+# Option 1: Load the decoder directly
+decoder_config = AutoConfig.from_pretrained("facebook/musicgen-melody").decoder
+decoder = MusicgenMelodyForCausalLM.from_pretrained("facebook/musicgen-melody", **decoder_config.to_dict())
 
->>> # Option 2: load the entire composite model, but only return the decoder
->>> decoder = MusicgenMelodyForConditionalGeneration.from_pretrained("facebook/musicgen-melody").decoder
+# Option 2: Access the decoder from the composite model
+model = MusicgenMelodyForConditionalGeneration.from_pretrained("facebook/musicgen-melody")
+decoder = model.decoder
 ```
```
Comment (stevhliu) [inline docs/source/en/model_doc/musicgen_melody.md:179]: Add a few more notes:

```md
- Ensure you're using a 32kHz checkpoint of the Encodec model because MusicGen was trained on it.
```

**Review by stevhliu — COMMENTED** [review]
Good start on this big doc!
